### PR TITLE
#4460 add volume support to ContainerClient

### PR DESCRIPTION
**Pull request by chia7712**
### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #4460)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-agent/src/main/scala/oharastream/ohara/agent/k8s/K8SClient.scala [hunk 1]
```diff
@@ -41,6 +41,20 @@ trait K8SClient extends ContainerClient {
   def nodes()(implicit executionContext: ExecutionContext): Future[Seq[K8SNodeReport]]
   def masterUrl: String
   def metricsUrl: Option[String]
+
+  // TODO: https://github.com/oharastream/ohara/issues/4460
```
Comment (chia7712) [inline ohara-agent/src/main/scala/oharastream/ohara/agent/k8s/K8SClient.scala:45]: @jackyoh 再麻煩你幫忙完成k8s的部分

File: ohara-it/src/test/scala/oharastream/ohara/it/collie/TestContainerClient.scala [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      TimeUnit.SECONDS.sleep(3)
+      log(name, Some(1)) should not include lastLine
+      log(name, Some(10)) should include(lastLine)
+    } finally Releasable.close(() => result(containerClient.forceRemove(name)))
+  }
+
+  @Test
+  def testVolume(): Unit = {
+    if (containerClient.isInstanceOf[K8SClient])
+      throw new AssumptionViolatedException("TODO: https://github.com/oharastream/ohara/issues/4460")
```
Comment (chia7712) [inline ohara-it/src/test/scala/oharastream/ohara/it/collie/TestContainerClient.scala:67]: @jackyoh 你完成k8s的部分後就可以把這段程式碼刪掉

File: ohara-agent/src/main/scala/oharastream/ohara/agent/k8s/K8SClient.scala (hunk 1 above)
Comment (jackyoh) [reply ohara-agent/src/main/scala/oharastream/ohara/agent/k8s/K8SClient.scala:45]: Ok, 沒問題。我再來研究看要如何實作

File: ohara-it/src/test/scala/oharastream/ohara/it/collie/TestContainerClient.scala (hunk 2 above)
Comment (jackyoh) [reply ohara-it/src/test/scala/oharastream/ohara/it/collie/TestContainerClient.scala:67]: 好，實作完我再來移除掉

**Comment by jackyoh** [comment]
retry collieIT

File: ohara-agent/src/main/scala/oharastream/ohara/agent/container/ContainerClient.scala [hunk 3]
```diff
@@ -53,29 +72,160 @@ trait ContainerClient extends Releasable {
...
+
+  /**
+    * @param name volumne name
+    * @return true if there is a deleted volume. Otherwise, false
+    */
+  def removeVolume(name: String)(implicit executionContext: ExecutionContext): Future[Unit]
 }
 
 object ContainerClient {
+  case class ContainerVolume(name: String, driver: String, path: String, nodeName: String)
```
Comment (jackyoh) [inline ohara-agent/src/main/scala/oharastream/ohara/agent/container/ContainerClient.scala:184]: 我來確認一下參數的目的：
name： volume 的名稱
driver ：我猜是指 local 還是 NFS
path:  node 的資料夾路徑
nodeName: 節點的名稱

**Comment by chia7712** [comment]
retry otherIT

File: ohara-agent/src/main/scala/oharastream/ohara/agent/container/ContainerClient.scala (hunk 3 above)
Comment (chia7712) [reply ohara-agent/src/main/scala/oharastream/ohara/agent/container/ContainerClient.scala:184]: you are totally right :)

Let me add enough comments for this class.

Comment (jackyoh) [reply ohara-agent/src/main/scala/oharastream/ohara/agent/container/ContainerClient.scala:184]: Thank you.

**Comment by chia7712** [comment]
unrelated error

**Pull request merged**